crates/cairo-coverage/src/data_loader/loaded_data.rs

```diff
@@ -1,65 +1,84 @@
+use anyhow::{Context, Result};
 use cairo_lang_sierra::debug_info::DebugInfo;
-use cairo_lang_sierra::program::{Program, ProgramArtifact, VersionedProgram};
+use cairo_lang_sierra_to_casm::compiler::CairoProgramDebugInfo;
 use camino::Utf8PathBuf;
 use derived_deref::Deref;
 use serde::de::DeserializeOwned;
 use std::collections::HashMap;
 use std::fs;
-use trace_data::CallTrace;
-
-type SourceSierraPath = String;
+use trace_data::{CallTrace, CallTraceNode, CasmLevelInfo};
 
 #[derive(Deref)]
-pub struct LoadedDataMap(HashMap<SourceSierraPath, LoadedData>);
+pub struct LoadedDataMap(HashMap<Utf8PathBuf, LoadedData>);
 
 pub struct LoadedData {
-    pub program: Program,
     pub debug_info: DebugInfo,
-    pub call_traces: Vec<CallTrace>,
+    pub casm_level_infos: Vec<CasmLevelInfo>,
+    pub casm_debug_info: CairoProgramDebugInfo,
 }
 
 impl LoadedDataMap {
-    pub fn load(call_trace_paths: &Vec<Utf8PathBuf>) -> Result<Self> {
-        let mut map: HashMap<SourceSierraPath, LoadedData> = HashMap::new();
-        for call_trace_path in call_trace_paths {
-            let call_trace: CallTrace = read_and_deserialize(call_trace_path)?;
+    pub fn load(call_trace_paths: &[Utf8PathBuf]) -> Result<Self> {
+        let execution_infos = call_trace_paths
+            .iter()
+            .map(read_and_deserialize)
+            .collect::<Result<Vec<_>>>()?
+            .into_iter()
+            .flat_map(load_nested_traces)
+            .filter_map(|call_trace| call_trace.cairo_execution_info)
+            .collect::<Vec<_>>();
 
-            let source_sierra_path = &call_trace
-                .cairo_execution_info
-                .as_ref()
-                .context("Missing key 'cairo_execution_info' in call trace. Perhaps you have outdated scarb?")?
-                .source_sierra_path;
+        // OPTIMIZATION:
+        // Group execution info by source Sierra path
+        // so that the same Sierra program does not need to be deserialized multiple times.
+        let execution_infos_by_sierra_path = execution_infos.into_iter().fold(
+            HashMap::new(),
+            |mut acc: HashMap<_, Vec<_>>, execution_info| {
+                acc.entry(execution_info.source_sierra_path)
+                    .or_default()
+                    .push(execution_info.casm_level_info);
+                acc
+            },
+        );
 
-            if let Some(loaded_data) = map.get_mut(&source_sierra_path.to_string()) {
-                loaded_data.call_traces.push(call_trace);
-            } else {
-                let VersionedProgram::V1 {
-                    program:
-                        ProgramArtifact {
-                            program,
+        Ok(Self(
+            execution_infos_by_sierra_path
+                .into_iter()
+                .map(|(source_sierra_path, casm_level_infos)| {
+                    read_and_deserialize::<SierraProgram>(&source_sierra_path)?
+                        .compile_and_get_debug_infos()
```

Comment:
general question: how does this affect the execution times? 
I wonder if there is anything we can do beforehand to have this or at least not to compile each execution (and - should we). Please let me know if this is something that was already discussed/though of beforehand.

Reply:
I know compiling in a loop might seem intimidating, but:

- We group the files in a hashmap beforehand so we only compile unique ones. This is a minimal number as each program needs to be compiled to CASM to get its debug info.
- I tested this on the OpenZeppelin repo, and it doesn't seem like a bottleneck (I know it is not a reliable benchmark).


The same strategy is used in the profiler, and from what I've heard, profiling is fast enough. So, these two projects have the same issue. We can try to work on it together if it seems like a major performance issue.

Reply:
The only edge case I can think of where we need to compile more than necessary is when `ContractClass` and `VersionedProgram` are the same program. But I haven't found a reliable way to distinguish between them.

Reply:
Hmm, ok thanks! Let's leave it then, I will try to look into this further in profiler and assess on my own. Will keep you updated :+1:

Reply:
`Snforge` already compiles to CASM using USC, so we could serialize it and pass it. Then, we wouldn't need to compile it, but we'll need to change the interface, and this implies that using this as a standalone binary won't be easy.

The same can be done in the profiler, so if we ever want to do that optimization, we can change this also there.

Reply:
It is not worth the effort, the gains would be really minimal (especially after https://github.com/software-mansion/universal-sierra-compiler/pull/57)